cub/device/dispatch/dispatch_histogram.cuh

```diff
@@ -243,14 +243,24 @@ public:
     struct ScaleTransform
     {
     private:
-      using CommonT = typename cuda::std::common_type<LevelT, SampleT>::type;
-      static_assert(cuda::std::is_convertible<CommonT, int>::value,
+      using CommonT = typename ::cuda::std::common_type<LevelT, SampleT>::type;
+      static_assert(::cuda::std::is_convertible<CommonT, int>::value,
                     "The common type of `LevelT` and `SampleT` must be "
                     "convertible to `int`.");
-      static_assert(cuda::std::is_trivially_copyable<CommonT>::value,
+      static_assert(::cuda::std::is_trivially_copyable<CommonT>::value,
                     "The common type of `LevelT` and `SampleT` must be "
                     "trivially copyable.");
 
+      // An arithmetic type that's used for bin computation of integral SampleT, guaranteed to not
+      // overflow for (max_level - min_level) * scale.fraction.bins. Since we drop invalid samples
+      // of less than min_level, (sample - min_level) is guaranteed to be non-negative. We use the
+      // rule: 2^l * 2^r = 2^(l + r) to determine a sufficiently large type to hold the
+      // multiplication result
+      using IntArithmeticT = cub::detail::conditional_t<       //
+        sizeof(SampleT) + sizeof(CommonT) <= sizeof(uint32_t), //
+        uint32_t,                                              //
+        uint64_t>;
```

Comment:
libcu++ recognizes `__{u,}int128_t`:
```cpp
#ifndef _LIBCUDACXX_HAS_NO_INT128
template <>          struct __libcpp_is_integral<__int128_t>         : public true_type {};
template <>          struct __libcpp_is_integral<__uint128_t>        : public true_type {};
#endif
```
Would this code trim the arithmetic value to `uint64`?

Reply:
Thanks, wasn't aware `::cuda::std::is_integral<T>` considers `__[u]int128_t` as `is_integral`. Per offline discussion, we agreed to: 
- apply promoting to wider integral types only up to `uint64_t`
- use 128-bit wide integer arithmetic only if `CommonT` is already 128 bits wide